pkg/apis/resource/validation/validation.go

```diff
@@ -0,0 +1,318 @@
+	allErrs := corevalidation.ValidateObjectMetaUpdate(&resourceClaim.ObjectMeta, &oldClaim.ObjectMeta, field.NewPath("metadata"))
+	fldPath := field.NewPath("status")
+	// The name might not be set yet.
+	if resourceClaim.Status.DriverName != "" {
+		allErrs = append(allErrs, validateResourceDriverName(resourceClaim.Status.DriverName, fldPath.Child("driverName"))...)
+	} else if resourceClaim.Status.Allocation != nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("driverName"), "must be specified when `allocation` is set"))
+	}
+
+	allErrs = append(allErrs, validateAllocationResult(resourceClaim.Status.Allocation, fldPath.Child("allocation"))...)
+	allErrs = append(allErrs, validateSliceIsASet(resourceClaim.Status.ReservedFor, resource.ResourceClaimReservedForMaxSize,
+		validateResourceClaimUserReference, fldPath.Child("reservedFor"))...)
+
+	// Now check for invariants that must be valid for a ResourceClaim.
+	if len(resourceClaim.Status.ReservedFor) > 0 {
+		if resourceClaim.Status.Allocation == nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "may not be specified when `allocated` is not set"))
+		} else {
+			if !resourceClaim.Status.Allocation.Shareable && len(resourceClaim.Status.ReservedFor) > 1 {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "may not be reserved more than once"))
+			}
+			// Items may be removed from ReservedFor while the claim is meant to be deallocated,
+			// but not added.
+			if resourceClaim.DeletionTimestamp != nil || resourceClaim.Status.DeallocationRequested {
+				oldSet := sets.New(oldClaim.Status.ReservedFor...)
+				newSet := sets.New(resourceClaim.Status.ReservedFor...)
+				newItems := newSet.Difference(oldSet)
+				if len(newItems) > 0 {
+					allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "new entries may not be added while the claim is meant to be deallocated"))
+				}
+			}
+		}
+	}
+
+	if !oldClaim.Status.DeallocationRequested &&
+		resourceClaim.Status.DeallocationRequested &&
+		len(resourceClaim.Status.ReservedFor) > 0 {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "deallocation cannot be requested while `reservedFor` is set"))
+	}
+
+	if resourceClaim.Status.Allocation == nil &&
+		resourceClaim.Status.DeallocationRequested {
+		// This combination is an error. We could just state that fact
+		// (i.e. "deallocationRequested: must not be set when
+		// `allocation` is nil"), but that is a confusing error message
+		// when the driver author tried to clear `allocation` and
+		// forgot about `deallocationRequested`. By looking at the
+		// change that is being made as part of the update we can
+		// provide a more helpful message.
+		if oldClaim.Status.Allocation != nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("deallocationRequested"), "must be cleared when `allocation` is unset"))
```

Comment:
FTR: In some cases we do this automatically in the registry logic, but that's more for end-user experience.  I don't think you need it here.

Reply:
I've simplified it to "`allocation` must be set when `deallocationRequested` is set", with the field path for "status".

That covers both scenarios and the user can figure out which field they modified incorrectly.